### PR TITLE
Update to latest Bazel rules

### PR DESCRIPTION
**Pull request by alexeagle**
- Updates required version of Bazel to latest
- Now use Bazel-managed deps (Bazel runs yarn install, user does not have to)
- Use fine-grained deps, rather than every action depending on all node_modules
- Global (ambient) typings for jasmine are now declared per-file so this typing does not pollute non-test code
- rules_typescript now comes purely as an npm package

### CODE REVIEW

File: src/tsickle.ts [hunk 1]
```diff
@@ -61,7 +62,7 @@ export function mergeEmitResults(emitResults: EmitResult[]): EmitResult {
   for (const er of emitResults) {
     diagnostics.push(...er.diagnostics);
     emitSkipped = emitSkipped || er.emitSkipped;
-    emittedFiles.push(...er.emittedFiles);
+    emittedFiles.push(...(er.emittedFiles as string[]));
```
Comment (evmar) [inline src/tsickle.ts]: I vaguely recall Rado has a patch for this, do you know why it fails?  I think we concluded to put a `!` on it instead of the cast.

File: test/e2e_clutz_dts_test.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+///<reference types="jasmine"/>
```
Comment (evmar) [inline test/e2e_clutz_dts_test.ts]: In every instance but this one you had a blank line after this import

File: BUILD [hunk 3]
```diff
@@ -13,12 +13,12 @@ filegroup(
...
+    ]) + [
         # The test suites run version-specific TypeScript compilers,
-        # but they all need the 'tslib' package to be available in
-        # a node_modules directory *above* the test/ directory.  So
-        # we must provide this extra directory as data to those tests.
-        "node_modules/tslib/*",
-    ]),
+        # but they all need the 'tslib' package to be available. So
+        # we must provide this extra filegroup as data to those tests.
+        "@npm//tslib:tslib__files",
```
Comment (evmar) [inline BUILD]: Sorry about the crazy hacks here

File: src/tsickle.ts (hunk 1 above)
Comment (alexeagle) [reply src/tsickle.ts]: it failed because the type of `er.emittedFiles` is `unknown`. Happy to take some other fix

File: BUILD (hunk 3 above)
Comment (alexeagle) [reply BUILD]: s'okay, I took a stab at removing the special handling for tslib but it was pretty easy to just update the location instead

File: test/e2e_clutz_dts_test.ts (hunk 2 above)
Comment (alexeagle) [reply test/e2e_clutz_dts_test.ts]: if the formatter doesn't format it then how can a human possibly do it? :man_shrugging:

File: src/tsickle.ts (hunk 1 above)
Comment (alexeagle) [reply src/tsickle.ts]: changed to non-null assertion operator, it works too

Comment (rkirov) [reply src/tsickle.ts]: see https://github.com/angular/tsickle/pull/990

File: test/test_support.ts [hunk 4]
```diff
@@ -37,7 +38,7 @@ function rootDir(): string {
  * tslib if we let it do its normal module resolution.  So instead we fix
  * the path here to the known path to the desired file.
  */
-const tslibPath = path.join(rootDir(), 'node_modules/tslib/tslib.d.ts');
+const tslibPath = path.join(rootDir(), '../npm/node_modules/tslib/tslib.d.ts');
```
Comment (rkirov) [inline test/test_support.ts]: This won't sync nicely into g3. Why is this change needed?

File: .circleci/config.yml [hunk 5]
```diff
@@ -32,5 +32,5 @@ jobs:
           key: *cache_key
       - run: bazel run @nodejs//:yarn
       - run: bazel build ...
-      - run: bazel test ...
+      - run: bazel test ... --nolegacy_external_runfiles
```
Comment (rkirov) [inline .circleci/config.yml]: any comment about this flag, so that we know when is it safe to remove?

File: src/tsickle.ts (hunk 1 above)
Comment (rkirov) [reply src/tsickle.ts]: Let's get this in first, and I will do the "proper fix" with #990 afterwards.

File: test/test_support.ts (hunk 4 above)
Comment (evmar) [reply test/test_support.ts]: We can fix the g3 sync after, no big deal

File: src/tsickle.ts (hunk 1 above)
Comment (alexeagle) [reply src/tsickle.ts]: I went ahead and applied your same change here too rather than check in an intermediate state

File: test/test_support.ts (hunk 4 above)
Comment (alexeagle) [reply test/test_support.ts]: This change is needed because the old path assumed that node_modules was a directory in our repository (meaning you manually installed it before running bazel)
Now that bazel manages the node modules, it lays them out in an external repository called npm

File: .circleci/config.yml (hunk 5 above)
Comment (alexeagle) [reply .circleci/config.yml]: Under current bazel, the external files are layed out in two places:
`runfiles_root/my_proj/external/npm/node_modules`
`runfiles_root/npm/node_modules`

They will eventually stop doing the first one. I added this flag to ensure that we are forwards compatible, opting into that breaking change. We could leave it here indefinitely.
Added a comment
